### ISSUE TITLE
[Feature Request] Support query in ZodObject

### ISSUE THREAD

**Issue by hanayashiki** [ecyrbe/zodios#406]
Right now if we want to support multiple queries, we need to write:

```ts
[
      {
        type: "Query",
        name: "limit",
        schema: z.number(),
      },
      {
        type: "Query",
        name: "offset",
        schema: z.number(),
      },
 ]
```

Which gets quite lengthy if you want to support a dozen of queries. In this way, each query costs 5 lines. What's more, this structure is hard to share with other parts in zod.

For example, I want to add a form for user to input the query, I might use the following schema:

```ts
z.object({
  limit: z.number(),
  offset: z.number()
})
```

There isn't a non-generic-intensive way to do that. Right now, I'm using the following generic function:

```ts
export type SchemaToQueryParametersResult<
  T extends ZodObject<ZodRawShape>,
  Keys extends readonly [...string[]],
> = {
  [Index in keyof Keys]: {
    name: Keys[Index];
    type: 'Query';
    description?: string;
    schema: T['shape'][Keys[Index]];
  };
};

/**
 * Convert a ZodObject to a corresponding tuple of query parameters.
 * Keys of the fields are required for now, because TypeScript does not support mapping object to tuple.
 * To use this in parameters, you should `...` the result to avoid
 * incorrect types.
 * @param t
 * @param keys
 * @returns
 */
export const schemaToQueryParameters = <
  T extends ZodObject<ZodRawShape>,
  Keys extends readonly [...string[]],
>(
  t: T,
  keys: Narrow<Keys>,
): SchemaToQueryParametersResult<T, Keys> => {
  const parameters = keys.map((key) => ({
    name: key,
    type: 'Query',
    description: (t.shape as any)[key].getFormMeta().description,
    schema: (t.shape as any)[key],
  }));
  return parameters as any;
};

type Try<A, B, C> = A extends B ? A : C;
type NarrowRaw<T> =
  | (T extends Function ? T : never)
  | (T extends string | number | bigint | boolean ? T : never)
  | (T extends [] ? [] : never)
  | {
      [K in keyof T]: NarrowRaw<T[K]>;
    };

type Narrow<T> = Try<T, [], NarrowRaw<T>>;
```

And the user should use

```ts
schemaToQueryParameters(
  z.object({
    limit: z.number(),
    offset: z.number()
  }),
  ['limit', 'offset'],
);
```

to convert the object into something like `[ type: 'Query', schema ..., type: 'Query', schema: ...]`.

The `Narrow<T>` is copied from your source code. I found no easy way to make typescript consider a user input `['limit', 'offset']` as tuple`["limit", "offset"]` instead of `string[]`.

This method also requires the user to repeat the keys as there are no good ways to convert an object into a tuple.

If I were to suggest, I might think the library should support a `{ type: 'Queries', schema: ZodObject }` type. It should be treated like multiple `{ type: 'Query', schema: ZodType }`. The library might merge all `Queries` and `Query` together, to form the input query type.

Thank you in advance.

**Comment by ecyrbe** [comment]
Hello, Thank you for the input, but this is not possible.
Indeed, zodios format is crafted to allow to create open api (swagger) from the api définition.
So this means that zodios needs to allow to declare description for each query outside of zod (because zodios v11 is independant of zod features)
But what can be done is create and expose an helper function like you did.
I think i can craft one that don't require to repeat thé entries.

**Comment by ecyrbe** [comment]
@hanayashiki here is a working [implementation](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwF4F84DMoRHA5EiAE1wG4BYAKEoFMAPSWOGATzGrgFUA7YCLgFQgBJLjGpQAztQDGMXlwA8AVx58AfHAC8cABSU4cFfLh0xXQhLgBDLszgB+XQGsAXIdVcAlFo0A3CMCEcG5c1L7ilN6m1OaWOq5wwFzo4omi4lKy8t6afgGE+g5pYpIycnyFIWHi5BQA9ABUDfoNcADm1PAANlYS8NRd1CAx8BDo7vItcAACYFZQVjjcxgC0nB5wY9a2TKzUElPTUJ1KUFyWADK9-YPDopvjRhUUcA11lCxscADinct8Vz6Cn+XA02hBghEJUy5UUhRBJloZgs2zsjh0OQ0CKq4SglA00ViukxaRSUDggJghUclMqcFCuNqlEazRerQAwnxcfAnlwmBAmEowINDnMFksNmsEVsbHZPvsADRwaQ2OAAI3YvIeyuwwroSukuogUiC0DgYCgoGAcnCuzYBzZM2OMFO5zg-CFg211Cs0gAFiZbiNLEkmH7NR4Wu8KHQGPB5et5IIPXrgR4lSmvYSUSonFwIAB3LgAbQAulo4GWwZX4ZGKOXs5ZiwzxKXqe7PdQ6RCIJnqApCgYDABRWjSLpKQj9kFK34wEGUtPyNRqBWDofFucL65L9RKgB0h77bZeQ7UTJj9Gg8b2cAAWkRBABFJTiZgABXmi06GQHp-4iLIpYSD7g+hAAPJqgAVmUCggWBABKVgFgAyn6VhsKuhQANLUMwliNpWh77n0lpcG0Zbltoo4wAssi7gIvadgoTh4Vs-DFrgAD6SRgEoMC4KWq7FsRpFJBRpZCWuFDVgghTFiIU60GkcCscwWy4fhpZuHJp4GFw35uJpEgKeYdClrUQ52tQbi4C+b6kOuU4SNIlpgLC9huGJ5GWUOLnhiAVhuBxuASOhbCCcWxmmUpkm+SgtQJTQV6MIa5zwAQhDPq+UAfl+wzQhWf4GABhHwUQkEwfR5WEEhqHhdQWEyXop4wMFkRuGB2Vvp+4o-pICj8GCGi6QYaV9Oa+X9ZY2iVWU+5qRIOgwCRDWePugVgDoql4T4uijUOBnDG4anSVZ8q2fZuW4GdQ7Oa5wDufIbgrWFGHUMWamlvu91ubCt0GP5QxBUwq3vZ9eGlrdKCeJ4vnOq6k19YVvSoolF7jfAxwSEoXTwNomXdblvXftCOggRA0FlDoB3hl0XQQG4IHeW0GK3QW0BdIQTP7lwSggBqUAYpQMNw0AA)

![image](https://user-images.githubusercontent.com/633115/229352984-32db6281-3ec3-4b22-aa08-11a559ddc0ae.png)

**Comment by hanayashiki** [comment]
Thank you, I'll definitely take a look!

**Comment by Zehir** [comment]
@hanayashiki you may be intrested by this package : https://github.com/thelinuxlich/zodios-api-shorthand

**Comment by hanayashiki** [comment]
> @hanayashiki you may be intrested by this package : https://github.com/thelinuxlich/zodios-api-shorthand

This package looks promising!

**Comment by ecyrbe** [comment]
New utility builders for params is now available on `v10.9.0`
See : https://twitter.com/ecyrbedev/status/1643022356502175745?t=UNuVBANKz3O-xCfN87HGVA&s=19
